### ISSUE TITLE
Stop ajax loading with `esc`

### ISSUE THREAD

**Issue by fregante** [sindresorhus/refined-github#2118]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/refined-github/issues/2118)

# Current

1. Click on the "Issues" tab above
2. Change your mind
3. Can't do anything, Issues keeps loading
4. When loaded, go back

# Desired

1. Click on the "Issues" tab above
2. Change your mind
3. Press `esc`, loading stops

⬆️ this is how regular pages work, but `pjax` doesn't listen to `esc`

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/refined-github/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by sindresorhus** [comment]
Sounds good.

**Comment by bfred-it** [comment]
`pjax` doesn't seem to offer a Cancel/Abort action.

To implement this you'd have to come up with some hack. Hope someone can think of something

**Comment by jerone** [comment]
It looks like the pjax on GitHub doesn't return the [`xhr`](https://github.com/defunkt/jquery-pjax/blob/master/jquery.pjax.js#L215) with [`pjax:beforeSend`](https://github.com/defunkt/jquery-pjax#events).

```js
$(document).on("pjax:beforeSend",function(){console.log(this, arguments);});
$("#js-repo-pjax-container").on("pjax:beforeSend",function(){console.log(this, arguments);});
document.getElementById("js-repo-pjax-container").addEventListener("pjax:beforeSend",function(){console.log(this, arguments);});
document.addEventListener("pjax:beforeSend",function(){console.log(this, arguments);});
```

**Comment by notlmn** [comment]
I'm not sure if https://github.com/defunkt/jquery-pjax is the one GitHub is using, because most of the events are missing from GitHub's JS files.

So, I came up with a hack using `AbortController`, that involves replacing `fetch` at global level.

``` js
const actualFetch = self.fetch; // Saving original fetch for later use
const controller = new AbortController();

// Custom fetch, replacing `AbortSignal` on request object
self.fetch = async (request) => {
	console.log(request);
	const newRequest = new Request(request, {signal: controller.signal}); // Create new request, with our own signal
	try {
		return actualFetch(newRequest);
	} catch (error) {
		console.error(error);
	}
};

// Handler to abort fetch
window.addEventListener('keydown', (event) => {
	if (event.key === 'Escape') {
		console.log('Aborting...');
		controller.abort();
	}
});
```

One catch with the above snippet is that after the fetch is aborted, the user is directly redirected to the target page, which is no the intended behavior, we need the user to stay on the current page.

I think we can manage this using as very clever combination of `mousedown`, `click`, and `mouseup` event handlers and using `stopPropagation()` if the click actually caused that aborted fetch (haven't tested that part though).

**Comment by bfred-it** [comment]
> replacing `fetch` at global level

Make sure that this actually works because last time I tried to do that in the content script it failed in Firefox, but now I'm testing in the console and it appears to work:


<img width="158" alt="Screenshot 2019-07-23 at 00 34 51" src="https://user-images.githubusercontent.com/1402241/61652269-be30d080-ace1-11e9-96dd-080fcaebb406.png">

However keep in mind that this replacing code has to be injected _unsafely_ in the main page like `resolve-conflicts` because content scripts are in a different realm.

**Comment by bfred-it** [comment]
Additionally, I think that GitHub will actually try to reload the page when the fetch fails, so `esc` would have to be pressed twice: once to stop the fetch and again to stop the reload.

**Comment by TheGuardianWolf** [comment]
Hey @fregante is this still an active issue?

**Comment by fregante** [comment]
It’s still active, but I don’t think it’s solvable. Do you want to try?

**Comment by kidonng** [comment]
Has anyone sent feedback to GitHub? If not I'm going to make one 😁

**Comment by TheGuardianWolf** [comment]
I can try, but in my firefox and chrome, I'm able to ESC out of loading issues, but only issues, is this the same for you guys?

**Comment by fregante** [comment]
> I'm able to ESC out of loading issues, but only issues

I think that's the regular ESC feature of the browser because they are not ajaxed pages. That's what this issue is about: ajaxed pages don't support ESC

**Comment by fregante** [comment]
> Has anyone sent feedback to GitHub? If not I'm going to make one 😁

Please do!

**Comment by TheGuardianWolf** [comment]
@fregante 

One hack coming right up, we can do this with webextensions like the following:

```typescript
const ajaxRequests = new Set<string>();
const cancelRequests = new Set<string>()

browser.commands.onCommand.addListener((command) => {
	if (command === 'stop-request') {
		cancelRequests.clear();
		Array.from(ajaxRequests).forEach((ajaxUrl) => {
			const requestedUrl = ajaxUrl.match(/^(.*)[?&]_pjax=.*$/i)[1];
			if (requestedUrl) {
				cancelRequests.add(ajaxUrl);
				cancelRequests.add(requestedUrl);
			}
		})
	}
})

browser.webRequest.onBeforeRequest.addListener((details) => {
	ajaxRequests.add(details.url);
}, { urls: ['https://github.com/*?*_pjax=*'], types: ['xmlhttprequest'] });

browser.webRequest.onHeadersReceived.addListener((details) => {
	ajaxRequests.delete(details.url);
}, { urls: ['https://github.com/*?*_pjax=*'], types: ['xmlhttprequest'] });

browser.webRequest.onHeadersReceived.addListener((details) => {
	if (cancelRequests.has(details.url)) {
		cancelRequests.delete(details.url);
		if (details.type === 'main_frame') {
			return { cancel: true };
		}
		else if (details.type === 'xmlhttprequest') {
			return { redirectUrl: details.originUrl };
		}
	}
}, { urls: ['https://github.com/*'], types: ['main_frame', 'xmlhttprequest'] }, ['blocking']);
```

**Comment by notlmn** [comment]
I don't believe we have access to `browser.webRequest`, if I'm not wrong isn't it a new permission we need to ask the user for?

**Comment by TheGuardianWolf** [comment]
Yeah, there are two permissions needed here `webRequest` and `webRequestBlocking`

**Comment by notlmn** [comment]
> Yeah, there are two permissions needed here `webRequest` and `webRequestBlocking`

Yeah, we wouldn't really want to add more permissions if we could do it any other way.

Also how is cancelling requests using `browser.webRequest` different from signalling using `AbortController`?

**Comment by TheGuardianWolf** [comment]
It's not different in cancellation, but the other method of the self redirect is possible through webRequests only.

The main advantage is not in cancelling the ajax request, as you've already shown that to be possible, but in cancelling the page load afterwards that pjax falls back on.

Was that the reason why you didn't continue with your fix? Or was there some other complication I'm not aware of?

**Comment by fregante** [comment]
> Yeah, there are two permissions needed here `webRequest` and `webRequestBlocking`

We can't add those 2 permissions by default, which means we'd have to have a UI for this specific feature to request them (`browser.permission.request`). For example it could append a message like

```
ESC won't cancel ajaxed loads, but `xyz-feature` can if you add extra permissions: [Add]
```

If your solution works, feel free to send a PR, keeping in mind that 👇

> some html content seems to lose their formatting
> content is rerendered

👆 these are probably deal-breakers. The point of this feature would be to avoid changing page; For example if you're not down leaving a review comment, a rerendering would still lose the comment.

**Comment by TheGuardianWolf** [comment]
That makes sense, then the cancellation method is more in line with what you want.

With that method and the issues, what I've seen is two issues:

* The profile page's username is not rendered properly after cancelling load
* DevTools no longer shows HTML after cancelling load

If these two issues are not acceptable then I guess it really can't be approached from outside of github code.

**Comment by fregante** [comment]
> * The profile page's username is not rendered properly after cancelling load on the profile page

This might be due to some _unload_ handler that we cannot prevent. We can consider these separately, like if it breaks profiles, we can disable it there, as long as it doesn't break _almost everywhere_


> * DevTools no longer shows HTML after cancelling load

Not a problem, GitHub users generally don't need to open the devtools.

Perhaps you can send an MVP PR to test it out

**Comment by TheGuardianWolf** [comment]
Wow, that pull request by @artusm is an elegant solution. I think that is your best bet.

To think that pjax event hooks were right under our noses 😆

**Comment by fregante** [comment]
From "impossible" to "done"

Thanks @artusm!

**Comment by ivan-aksamentov** [comment]
Excuse me for touching the old issue, but how can I remove these `_pjax` entries in my urls?

```
_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D"http%3A%2F%2Fschema.org%2FSoftwareSourceCode"%5D%20main%2C%20%5Bdata-pjax-container%5D
```

These long additions pollute the links to code fragments, making them hard to read, and Slack does not always understand them either. 

I absolutely don't care about the problem described in this issue. Is this a feature of GitHub Refined or Github in general? Is there a toggle?
